Refactor headerInfo

MoonBow-1 opened the pull request:
Separate PR from Pr https://github.com/teragrep/cfe_16/pull/15 to make the current PR much more smaller in scope

Closes #42

- One test will fail, since that is not supposed to be fixed by this PR
- Makes HeaderInfo immutable as well
- Will require changes to the PRs that follow this, since the Converter class will be removed in favor of an "event" making itself to a SyslogEvent

MoonBow-1 committed:
Bring newer HeaderInfo from `fix-mocking` branch

MoonBow-1 committed:
Improve RequestHandler by using proper ctor

MoonBow-1 committed:
Use the HeaderInfo class in Converter

MoonBow-1 committed:
Add test file from `fix-mocking` branch

MoonBow-1 committed:
Make HeaderInfo immutable

MoonBow-1 committed:
Add equals tests for HeaderInfo

eemhu commented on src/main/java/com/teragrep/cfe_16/RequestHandler.java:
```diff
@@ -58,33 +58,48 @@
+            forwardedFor = xForwardedFor;
+        }
+        else {
+            forwardedFor = null;
```
should remove null assignments

MoonBow-1 replied:
Fixed in commit 9cf6ff4a

eemhu commented on src/main/java/com/teragrep/cfe_16/bo/HeaderInfo.java:
```diff
@@ -45,44 +45,57 @@
+    private final String xForwardedHost;
+    private final String xForwardedProto;
 
     public HeaderInfo(String xForwardedFor, String xForwardedHost, String xForwardedProto) {
```
the ctor args can also be final

MoonBow-1 replied:
Fixed in commit d20c6e97

eemhu commented on src/main/java/com/teragrep/cfe_16/RequestHandler.java:
```diff
@@ -58,33 +58,48 @@
+        this.request = request;
+    }
+
+    public HeaderInfo createHeaderInfoObject() {
```
this could be renamed -> `headerInfo()`

MoonBow-1 replied:
Fixed in commit 2eae9d11

eemhu commented on src/test/java/com/teragrep/cfe_16/bo/HeaderInfoTest.java:
```diff
@@ -0,0 +1,85 @@
+class HeaderInfoTest {
+
+    @Test
+    @DisplayName("asSDElement() uses all three parameters if they are not null")
```
also test with non present values ?

MoonBow-1 replied:
Test added in commit fb8f59d3

MoonBow-1 committed:
Set HeaderInfo ctor args to be final

MoonBow-1 committed:
Rename RequestHandler.createHeaderInfoObject() to RequestHandler.headerInfo()

MoonBow-1 committed:
Test that HeaderInfo.asSDElement() ignores null values

MoonBow-1 committed:
Remove RequestHandler.java since it only had one method

- Code moved to HECServiceImpl.sendEvents

Laukkala commented on src/test/java/com/teragrep/cfe_16/it/ServiceAndEventManagerIT.java:
```diff
@@ -145,7 +145,7 @@ public static void cleanup() throws InterruptedException {
     private EventManager eventManager;
 
-    private HeaderInfo headerInfo = new HeaderInfo();
+    private HeaderInfo headerInfo = new HeaderInfo(null, null, null);
```
Is there any way to avoid assigning nulls to HeaderInfo's constructor parameters, or is supporting nulls as parameters required due to some other component? Currently the only place where these nulls can be accessed is the asSDElement() method where they are checked, but if we later change them to be accessible to some other object, they might return nulls.

MoonBow-1 replied:
Due to the fact that CFE_16 handles HTTP requests, there is a real possibility that the
```java
public ObjectNode sendEvents(HttpServletRequest request, String channel, String eventInJson) {
...
}
```
parameter `request` does not have the HTTP headers that the `HeaderInfo` would like to have.

---

Current implementation is quite simple, where null values are expected and header values are used as they are:
```java
final String xForwardedFor = request.getHeader("X-Forwarded-For");
final String xForwardedHost = request.getHeader("X-Forwarded-Host");
final String xForwardedProto = request.getHeader("X-Forwarded-Proto");

final HeaderInfo headerInfo = new HeaderInfo(xForwardedFor, xForwardedHost, xForwardedProto);
```

---

I think the architect's opinion in required in this case

kortemik replied:
Stub-object is always a good way to avoid nulls and quite easy to insert while refactoring. 

```java
// in class definition:
private final XForwardedForStub xForwardedForStub; // stub is immutable and can be reused always

// in cTor
this.xForwardedForStub = new XForwardedForStub(); 

// in method
final XForwardedFor xForwardedFor;
if (request.getHeader("X-Forwarded-For") == null) {
  xForwardedFor = xForwardedForStub; // uses the immutable stub
}
else {
  xForwardedFor = new xForwardedForImpl(request.getHeader("X-Forwarded-For"));
}
```

@eemhu mind writing an how-to about this in fkb?

MoonBow-1 replied:
Will implement the three stub objects and remove null checks from the `HeaderInfo.asSDElement()`

kortemik replied:
created sac_01/issue/207 for @eemhu

MoonBow-1 committed:
Create objects for HTTP headers

- Includes stubs

MoonBow-1 committed:
Implement header classes to HeaderInfo

- Adds Stubable interface
- Extends the header interfaces from the Stubable interface

MoonBow-1 committed:
Remove unused imports from tests

kortemik commented on src/main/java/com/teragrep/cfe_16/bo/XForwardedFor.java:
```diff
@@ -0,0 +1,53 @@
+
+public interface XForwardedFor extends Stubable {
+
+    public abstract String value() throws UnsupportedOperationException;
```
UnsupportedOperationException is not a checked exception and therefore does not need to be explicitly mentioned in the interface signature

kortemik commented on src/main/java/com/teragrep/cfe_16/bo/XForwardedHost.java:
```diff
@@ -0,0 +1,53 @@
+
+public interface XForwardedHost extends Stubable {
+
+    public abstract String value() throws UnsupportedOperationException;
```
UnsupportedOperationException is not a checked exception and therefore does not need to be explicitly mentioned in the interface signature

kortemik commented on src/main/java/com/teragrep/cfe_16/bo/XForwardedProto.java:
```diff
@@ -0,0 +1,53 @@
+
+public interface XForwardedProto extends Stubable {
+
+    public abstract String value() throws UnsupportedOperationException;
```
UnsupportedOperationException is not a checked exception and therefore does not need to be explicitly mentioned in the interface signature

MoonBow-1 committed:
Remove unnecessary throws for unchecked exceptions

MoonBow-1 committed:
Remove unnecessary throw for unchecked exception in XForwardedFor
